### PR TITLE
feat(ssa): Replace values which have previously been constrained with simplified value

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

This PR adds a mapping for `ValueId`s which are constrained to be equal within the current block. When resolving the inputs of instructions, we can then alternate between the standard resolution and checking for an constrained equivalence in this mapping.

This optimization is especially useful for situations where we want to perform slice indexing such as in https://github.com/AztecProtocol/aztec-packages/blob/94053e44f4d2a702fe9066bfff3cdd35e6d1b645/yarn-project/noir-libs/noir-aztec/src/messaging/l1_to_l2_message.nr#L47-L65. Rather than checking the slice length on each access, we can assert the length of the slice once and skip all of the other checks.

We must keep a separate cache rather than inserting these equivalences into the `dfg` as otherwise we'll apply these simplifications before (and inside) the constrain statement which enforces it.

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

The bytecode for a number of programs has increased due to previous PRs which didn't update the committed bytecode. The changes from this optimzation are in e4c9634bf5a8c32dc7c71d246811f917b0a63a22

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by jfecher — COMMENTED** [review]
Since the scope of the constant_folding pass has expanded, can we add to its doc comment the new optimizations it is expected to do?

**Comment by TomAFrench** [comment]
This is ready for a re-review now. CI failure is unrelated to this PR.

File: crates/noirc_evaluator/src/ssa/opt/constant_folding.rs [hunk 1]
```diff
@@ -105,20 +110,46 @@ impl Context {
...
+    ) -> Instruction {
         let instruction = dfg[instruction_id].clone();
 
+        // Alternate between resolving `value_id` in the `dfg` and checking to see if the resolved value
+        // has been constrained to be equal to some simpler value in the current block.
+        //
+        // This allows us to reach a stable final `ValueId` for each instruction input as we add more
+        // constraints to the cache.
+        fn resolve_cache(
+            dfg_resolver: impl Fn(ValueId) -> ValueId,
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/opt/constant_folding.rs:138]: Why does this need to be a parameter? It looks like its only value is `|v| dfg.resolve(v)`

File: crates/noirc_evaluator/src/ssa/opt/constant_folding.rs [hunk 2]
```diff
@@ -321,4 +393,55 @@ mod test {
...
         assert_eq!(instruction, &Instruction::Cast(ValueId::test_new(0), Type::unsigned(32)));
     }
+
+    #[test]
+    fn constrained_value_replacement() {
+        // fn main f0 {
+        //   b0(v0: Field):
+        //     constrain v0 10
+        //     v1 = add v0, Field 1
+        //     constrain v1 11
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/opt/constant_folding.rs:403]:
```suggestion
        //     constrain v0 == Field 10
        //     v1 = add v0, Field 1
        //     constrain v1 == Field 11
```

File: crates/noirc_evaluator/src/ssa/opt/constant_folding.rs (hunk 1 above)
Comment (TomAFrench) [reply crates/noirc_evaluator/src/ssa/opt/constant_folding.rs:138]: We need a parameter here as `resolve_cache` can't capture its environment and we can't make it a closure as we need to recurse.

I've switched this over to just pass in `dfg` directly however.

File: crates/noirc_evaluator/src/ssa/opt/constant_folding.rs [hunk 3]
```diff
@@ -321,4 +393,55 @@ mod test {
...
+        let mut ssa = builder.finish();
+        let main = ssa.main_mut();
+        let instructions = main.dfg[main.entry_block()].instructions();
+        assert_eq!(instructions.len(), 3);
+
+        // Expected output:
+        //
+        // fn main f0 {
+        //   b0(v0: Field):
+        //     constrain v0 10
```
Comment (TomAFrench) [inline crates/noirc_evaluator/src/ssa/opt/constant_folding.rs:431]:
```suggestion
        //     constrain v0 == Field 10
```

**Review by jfecher — APPROVED** [review]
LGTM. I'm assuming the test is still failing though.

**Pull request merged**